pkg/http/listener.go

```diff
@@ -0,0 +1,326 @@
+					bufconn.RemoteAddr(), bufconn.LocalAddr())
+			}
+			bufconn.Close()
+			return
+		}
+
+		// Return bufconn if read data is a valid HTTP method.
+		tokens := strings.SplitN(string(data), " ", 2)
+		if isHTTPMethod(tokens[0]) {
+			if listener.tlsConfig == nil {
+				send(acceptResult{bufconn, nil}, doneCh)
+			} else {
+				// As TLS is configured and we got plain text HTTP request,
+				// return 403 (forbidden) error.
+				bufconn.Write(sslRequiredErrMsg)
+				bufconn.Close()
+			}
+			return
+		}
+
+		if listener.tlsConfig != nil {
+			// As the listener is configured with TLS, try to do TLS handshake, drop the connection if it fails.
+			tlsConn := tls.Server(bufconn, listener.tlsConfig)
+			if err := tlsConn.Handshake(); err != nil {
+				if listener.errorLogFunc != nil {
+					listener.errorLogFunc(err,
+						"TLS handshake failed with new connection %s at server %s",
+						bufconn.RemoteAddr(), bufconn.LocalAddr())
+				}
+				bufconn.Close()
+				return
+			}
+
+			// Check whether the connection contains HTTP request or not.
+			bufconn = newBufConn(tlsConn, listener.readTimeout, listener.writeTimeout,
+				listener.updateBytesReadFunc, listener.updateBytesWrittenFunc)
+
+			// Peek bytes of maximum length of all HTTP methods.
+			data, err := bufconn.Peek(methodMaxLen)
+			if err != nil {
+				if listener.errorLogFunc != nil {
+					listener.errorLogFunc(err,
+						"Error in reading from new TLS connection %s at server %s",
+						bufconn.RemoteAddr(), bufconn.LocalAddr())
+				}
+				bufconn.Close()
+				return
+			}
+
+			// Return bufconn if read data is a valid HTTP method.
+			tokens := strings.SplitN(string(data), " ", 2)
+			if isHTTPMethod(tokens[0]) {
+				send(acceptResult{bufconn, nil}, doneCh)
+				return
+			}
+		}
+
+		if listener.errorLogFunc != nil {
+			listener.errorLogFunc(errors.New("junk message"),
+				"Received non-HTTP message from new connection %s at server %s",
+				bufconn.RemoteAddr(), bufconn.LocalAddr())
+		}
+
+		bufconn.Close()
+		return
+	}
+
+	// Closure to handle new connections till done channel is not closed.
+	handleConnection := func(tcpListener *net.TCPListener, doneCh <-chan struct{}) {
+		for {
+			tcpConn, err := tcpListener.AcceptTCP()
+			if err != nil {
+				// Returns when send fails.
+				if !send(acceptResult{nil, err}, doneCh) {
+					return
+				}
+			} else {
+				go acceptTCP(tcpConn, doneCh)
+			}
+		}
+	}
+
+	// Start separate goroutine for each TCP listener to handle connection.
+	for _, tcpListener := range listener.tcpListeners {
+		doneCh := make(chan struct{})
+		go handleConnection(tcpListener, doneCh)
+		listener.doneChs = append(listener.doneChs, doneCh)
+	}
+}
+
+// Accept - reads from httpListener.acceptCh for one of previously accepted TCP connection and returns the same.
+func (listener *httpListener) Accept() (conn net.Conn, err error) {
+	result, ok := <-listener.acceptCh
+	if ok {
+		return result.conn, result.err
+	}
+
+	return nil, io.EOF
```

Comment:
is Accept supposed to give io.EOF? this is not an I/O operation... perhaps you should use a better error here.